core/services/ocr2/plugins/ocr2keeper/evm21/mercury/v02/request.go

```diff
@@ -0,0 +1,188 @@
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm21/mercury"
+)
+
+const (
+	mercuryPathV02      = "/client?" // only used to access mercury v0.2 server
+	retryDelay          = 500 * time.Millisecond
+	totalAttempt        = 3
+	contentTypeHeader   = "Content-Type"
+	authorizationHeader = "Authorization"
+	timestampHeader     = "X-Authorization-Timestamp"
+	signatureHeader     = "X-Authorization-Signature-SHA256"
+)
+
+type MercuryV02Response struct {
+	ChainlinkBlob string `json:"chainlinkBlob"`
+}
+
+type client struct {
+	mercuryConfig mercury.MercuryConfigProvider
+	httpClient    mercury.HttpClient
+	lggr          logger.Logger
+}
+
+func NewClient(mercuryConfig mercury.MercuryConfigProvider, httpClient mercury.HttpClient, lggr logger.Logger) *client {
+	return &client{
+		mercuryConfig: mercuryConfig,
+		httpClient:    httpClient,
+		lggr:          lggr,
+	}
+}
+
+func (c *client) DoRequest(ctx context.Context, streamsLookup *mercury.StreamsLookup, pluginRetryKey string) (mercury.MercuryUpkeepState, mercury.MercuryUpkeepFailureReason, [][]byte, bool, time.Duration, error) {
+	resultLen := len(streamsLookup.Feeds)
+	ch := make(chan mercury.MercuryData, resultLen)
+	if len(streamsLookup.Feeds) == 0 {
+		return mercury.NoPipelineError, mercury.MercuryUpkeepFailureReasonInvalidRevertDataInput, [][]byte{}, false, 0 * time.Second, fmt.Errorf("invalid revert data input: feed param key %s, time param key %s, feeds %s", streamsLookup.FeedParamKey, streamsLookup.TimeParamKey, streamsLookup.Feeds)
+	}
+	for i := range streamsLookup.Feeds {
+		go c.singleFeedRequest(ctx, ch, i, streamsLookup)
+	}
+
+	var reqErr error
+	var retryInterval time.Duration
+	results := make([][]byte, len(streamsLookup.Feeds))
+	retryable := true
+	allSuccess := true
+	// in v0.2, use the last execution error as the state, if no execution errors, state will be no error
+	state := mercury.NoPipelineError
+	for i := 0; i < resultLen; i++ {
+		m := <-ch
+		if m.Error != nil {
+			reqErr = errors.Join(reqErr, m.Error)
+			retryable = retryable && m.Retryable
+			allSuccess = false
+			if m.State != mercury.NoPipelineError {
+				state = mercury.MercuryUpkeepState(m.State)
+			}
+			continue
+		}
+		results[m.Index] = m.Bytes[0]
+	}
+	if retryable && !allSuccess {
+		retryInterval = mercury.CalculateRetryConfigFn(pluginRetryKey, c.mercuryConfig)
+	}
+	// only retry when not all successful AND none are not retryable
+	return state, mercury.MercuryUpkeepFailureReasonNone, results, retryable && !allSuccess, retryInterval, reqErr
+}
+
+func (c *client) singleFeedRequest(ctx context.Context, ch chan<- mercury.MercuryData, index int, sl *mercury.StreamsLookup) {
+	var httpRequest *http.Request
+	var err error
```

Comment:
Can we drop these two vars and change line 98 to `httpRequest, err := http.NewRequestWithContext(ctx, http.MethodGet, reqUrl, nil)`, i.e. declare them just when they're needed?

Reply:
this change is bc of the shadowing discussion we had before. declaring err here to avoid shadowing. I like the idea of having variables close to their usage, so no strong opinion on either approach.